source/develop/release-management/features/virt/dedicated-cpu.md

```diff
@@ -0,0 +1,170 @@
+vulnerabilities) after finished migration or when restoring VM from snapshot.
+
+## Limitations
+
+See the *Security* section.
+
+Currently, when some CPUs are already dedicated to a VM they cannot be evicted
+and provided to another VM. Notably this means:
+
+  * it may be impossible to start a VMs with CPU pinning on that host because
+    the requested CPUs are already occupied;
+
+  * it may be impossible to start a VM with dedicated CPUs even though there
+    seems to be enough space in terms of number of used vs. free CPUs.
+
+To workaround this limitation one can start VMs with CPU pinning before VMs
+with dedicated CPU policy. And more generally, the VMs with dedicated CPUs can
+be manually migrated to another host to free the needed CPUs and then possibly
+migrated back.
+
+*TODO*: Tips how to identify candidates for migration?  virsh? Something in UI?
+
+
+## User Experience
+
+## Changes to Engine and Vdsm
+
+The high level overview of the workflow in Engine and Vdsm is following:
+
+1.  When host switches to *UP* state engine queries it to find out which pCPUs
+    are dedicated through a policy or pinned. This is necessary for engine to
+    get a good picture about CPU usage so that it can schedule new VMs
+    properly.
+
+2.  On VM start/migration/etc. Engine consults the policy for dedicating CPUs
+    and schedules the VM on host with enough free CPUs. Engine determines the
+    mapping between physical and virtual CPUs.
+
+3.  Engine produces domain XML that contains properly populated `<vcpupin>`
+    element (similarly to CPU pinning) and stores the policy name in VM
+    metadata.
+
+4.  Vdsm starts the VM and analyzes which CPUs were dedicated. Vdsm updates the
+    pool of shared CPUs and re-configures VMs without any CPU pinning or
+    dedicated CPUs to run on the new set of CPUs from shared pool.
+
+5.  When the VM with CPU pinning or dedicated CPUs is being destroyed, Vdsm
+    updates the pool of shared CPUs. Vdsm then re-configures VMs without any
+    CPU pinning or dedicated CPUs to run on the new set of CPUs from shared
+    pool.
+
+Several changes to the internal API are required. To accomplish step 1. new API
+verb `Host.cpuInfo` needs to be defined that will provide:
+
+  * list of pCPUs dedicated to VMs through CPU pinning or policy
+  * CPU indices and topology information how are sockets, cores and threads
+    organized
+
+Further, Engine will add a new entry into VM metadata called `cpuPolicy` that
+will contain the requested CPU policy. And finally, migration parameters in
+`VM.migrate` will contain a field `cpus` that will contain the CPU mapping for
+the VM on the destination host. This is necessary so that the source VDSM can
+provide a correct domain XML to the destination host.
+
+*TODO*: In what format? Can we use `<vcpupin>` element content or should we use
+something more concise? Alternatively, we can also let destination VDSM parse
```

Comment:
I'd say <vcpupin> format is OK -- metadata is unfortunately no small thing anyway and reusing <vcpupin> format would keep things easy to understand.

Reply:
I see. I think using ``<vcpuping>`` would still be OK, in accordance with the general approach of preferably providing XML snippets from Engine directly.